source/common/config/delta_subscription_state.cc

```diff
@@ -81,13 +159,29 @@ bool DeltaSubscriptionState::isHeartbeatResponse(
       !Runtime::runtimeFeatureEnabled("envoy.reloadable_features.vhds_heartbeats")) {
     return false;
   }
-  const auto itr = resource_state_.find(resource.name());
-  if (itr == resource_state_.end()) {
+  if (resource.has_resource()) {
     return false;
   }
 
-  return !resource.has_resource() && !itr->second.waitingForServer() &&
-         resource.version() == itr->second.version();
+  if (const auto maybe_resource = getRequestedResourceState(resource.name());
+      maybe_resource.has_value()) {
+    return !maybe_resource->isWaitingForServer() && resource.version() == maybe_resource->version();
+  }
+
+  if (const auto itr = wildcard_resource_state_.find(resource.name());
+      itr != wildcard_resource_state_.end()) {
+    return resource.version() == itr->second;
+  }
+
+  if (const auto itr = ambiguous_resource_state_.find(resource.name());
```

Comment:
Maybe refactor into a version lookup method?

Reply:
Uh, maybe I am being dense now, but could you expand on it a bit? :) Did you mean having a function like `bool hasResourceWithVersion(std::string_view name, std::string_view version)` or a couple of functions like `bool has{Requested,Ambiguous,Wildcard}ResourceWithVersion(std::string_view name, std::string_view version)`? Thanks.